aws/data_source_aws_ec2_transit_gateway_vpn_attachment.go

```diff
@@ -19,34 +19,56 @@ func dataSourceAwsEc2TransitGatewayVpnAttachment() *schema.Resource {
 			"tags": tagsSchemaComputed(),
 			"transit_gateway_id": {
 				Type:     schema.TypeString,
-				Required: true,
+				Optional: true,
 			},
 			"vpn_connection_id": {
 				Type:     schema.TypeString,
-				Required: true,
+				Optional: true,
 			},
+			"filter": dataSourceFiltersSchema(),
 		},
 	}
 }
 
 func dataSourceAwsEc2TransitGatewayVpnAttachmentRead(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).ec2conn
 
+	filters, filtersOk := d.GetOk("filter")
+	tags, tagsOk := d.GetOk("tags")
+	connectionId, connectionIdOk := d.GetOk("vpn_connection_id")
+	transitGatewayId, transitGatewayIdOk := d.GetOk("transit_gateway_id")
+
 	input := &ec2.DescribeTransitGatewayAttachmentsInput{
 		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("resource-type"),
+				Values: []*string{aws.String(ec2.TransitGatewayAttachmentResourceTypeVpn)},
+			},
+		},
+	}
+
+	if filtersOk {
+		input.Filters = append(input.Filters, buildAwsDataSourceFilters(filters.(*schema.Set))...)
+	}
+	if tagsOk {
+		input.Filters = append(input.Filters, ec2TagFiltersFromMap(tags.(map[string]interface{}))...)
+	}
+	// to preserve original functionality
+	if connectionIdOk && transitGatewayIdOk {
+		input.Filters = []*ec2.Filter{
 			{
 				Name:   aws.String("resource-id"),
-				Values: []*string{aws.String(d.Get("vpn_connection_id").(string))},
+				Values: []*string{aws.String(connectionId.(string))},
 			},
 			{
 				Name:   aws.String("resource-type"),
 				Values: []*string{aws.String(ec2.TransitGatewayAttachmentResourceTypeVpn)},
 			},
 			{
 				Name:   aws.String("transit-gateway-id"),
-				Values: []*string{aws.String(d.Get("transit_gateway_id").(string))},
+				Values: []*string{aws.String(transitGatewayId.(string))},
 			},
-		},
+		}
 	}
```

Comment:
This looks like it could cause the unexpected consequence of removing configured `filter` options by overwriting the `input.Filters` parameter when all three are provided. I would recommend switching this to `append()` instead. 👍 

```go
	if connectionIdOk {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name:   aws.String("resource-id"),
			Values: []*string{aws.String(connectionId.(string))},
		})
	}

	if transitGatewayIdOk {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name:   aws.String("transit-gateway-id"),
			Values: []*string{aws.String(transitGatewayId.(string))},
		})
	}
```

Reply:
gotcha

Reply:
done,

--- PASS: TestAccAWSEc2TransitGatewayVpnAttachmentDataSource_TransitGatewayIdAndVpnConnectionId (501.92s)
--- PASS: TestAccAWSEc2TransitGatewayVpnAttachmentDataSource_filter (624.90s)